### PR TITLE
Solution

### PR DESCRIPTION
**Pull request by kateryna3k**
Screenshot 2022-11-02 at 14 28 40
Screenshot 2022-11-02 at 14 28 57
Screenshot 2022-11-02 at 14 29 05
Screenshot 2022-11-02 at 14 29 08
Screenshot 2022-11-02 at 14 29 02

### CODE REVIEW

File: templates/includes/sidebar.html [hunk 1]
```diff
@@ -1,6 +1,6 @@
-<ul class="sidebar-nav">
-  <li><a href="#">Home page</a></li>
-  <li><a href="#">Manufacturers</a></li>
-  <li><a href="#">Cars</a></li>
-  <li><a href="#">Drivers</a></li>
+<ul class="sidebar-nav" style="list-style-type: none">
```
Comment (shurushku) [inline templates/includes/sidebar.html:1]: You can add this in static/css. I think this best practice

**Review by shurushku — APPROVED** [review]

File: taxi/urls.py [hunk 2]
```diff
@@ -1,9 +1,25 @@
...
+                    ManufacturerListView,
+                    CarListView, DriverListView,
+                    CarDetailView, DriverDetailView)
 
 urlpatterns = [
     path("", index, name="index"),
+    path(
+        "manufacturers/",
+        ManufacturerListView.as_view(),
+        name="manufacturer-list"),
```
Comment (TPolina) [inline taxi/urls.py:13]: the close bracket should be from the next line

File: taxi/urls.py [hunk 3]
```diff
@@ -1,9 +1,25 @@
...
+    path("car/", CarListView.as_view(), name="car-list"),
+    path("driver/", DriverListView.as_view(), name="driver-list"),
+    path("car/<int:pk>/", CarDetailView.as_view(), name="car-detail"),
+    path(
+        "driver/<int:pk>",
+        DriverDetailView.as_view(),
+        name="driver-detail"
+    ),
```
Comment (TPolina) [inline taxi/urls.py:21]: common practice is to use plural forms:
cars/ - car list view;
cars/pk/ - car detail view;
drivers/ - driver list view;
drivers/pk/ - driver detail view

File: taxi/urls.py [hunk 4]
```diff
@@ -1,9 +1,25 @@
...
+
```
Comment (TPolina) [inline taxi/urls.py:22]: unnecessary blank line

File: taxi/views.py [hunk 5]
```diff
@@ -17,3 +18,31 @@ def index(request):
...
+
+
+class DriverListView(generic.ListView):
+    model = Driver
+    paginate_by = 5
+
+
+class DriverDetailView(generic.DetailView):
+    model = Driver
+    queryset = Driver.objects.all().prefetch_related("cars")
```
Comment (TPolina) [inline taxi/views.py:48]: you need to optimize getting manufacturers of the cars as well:
https://docs.djangoproject.com/en/4.1/ref/models/querysets/#prefetch-related

File: templates/includes/sidebar.html [hunk 6]
```diff
@@ -1,6 +1,6 @@
...
+  <li><a href="{% url "taxi:index" %}">Home page</a></li>
+  <li><a href="{% url "taxi:manufacturer-list" %}">Manufacturers</a></li>
+  <li><a href="{% url "taxi:car-list" %}">Cars</a></li>
+  <li><a href="{% url "taxi:driver-list" %}">Drivers</a></li>
```
Comment (TPolina) [inline templates/includes/sidebar.html:5]: use single quotes inside double

File: templates/taxi/car_detail.html [hunk 7]
```diff
@@ -0,0 +1,10 @@
+{% extends "base.html" %}
+{% block content %}
+  <p><strong>Car manufacturer: </strong>{{ car.manufacturer.name }}, {{ car.manufacturer.country }}</p>
```
Comment (TPolina) [inline templates/taxi/car_detail.html:3]: add info about car model

File: templates/taxi/car_detail.html [hunk 8]
```diff
@@ -0,0 +1,10 @@
...
+  <hr>
+  <ul style="list-style-type:none;">
+  {% for driver in car.drivers.all %}
+    <li>First name: {{ driver.first_name }}, Last name: {{ driver.last_name }}</li>
+  {% endfor %}
```
Comment (TPolina) [inline templates/taxi/car_detail.html:8]: add line `Drivers:` before info of each one
And also it's better to add a separate phrase in case there are no drivers

File: templates/taxi/driver_detail.html [hunk 9]
```diff
@@ -0,0 +1,12 @@
+{% extends "base.html" %}
+{% block content %}
+  <p><strong>First name: </strong>{{ driver.first_name }}, <strong>Last name: </strong>{{ driver.last_name }},
```
Comment (TPolina) [inline templates/taxi/driver_detail.html:3]: add username

File: templates/taxi/driver_detail.html [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+  <strong>License number: </strong>{{ driver.license_number }}
+  </p>
+  {% for car in driver.cars.all %}
+    <hr>
+    <p><strong>Id</strong>: {{ car.id }},
+    <strong>Car model</strong>: {{ car.model }}
+    </p>
+  {% endfor %}
```
Comment (TPolina) [inline templates/taxi/driver_detail.html:11]: add a `Car:` phrase before info about each. It is better to make info about each car as a list item

File: templates/taxi/driver_list.html [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+  {% if driver_list %}
+    <ul style="list-style-type:none;">
+    {% for driver in driver_list %}
+      <li>
+        <a href="{% url "taxi:driver-detail" pk=driver.id %}">{{ driver.username }}</a>
+        <strong>First name:</strong> {{ driver.first_name }},
+        <strong>Last name:</strong> {{ driver.last_name }}
+      </li>
+    {% endfor %}
+    </ul>
```
Comment (TPolina) [inline templates/taxi/driver_list.html:13]: add phrase if there are no drivers

File: templates/taxi/manufacturer_list.html [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+
+{% block content %}
+  <h1>Manufacturers list</h1>
+  {% if manufacturer_list %}
+  <ul style="list-style-type:none;">
+    {% for manufacturer in manufacturer_list %}
+      <li><strong>Manufacturer name:</strong> {{ manufacturer.name }},
+        <strong>Manufacturer country:</strong> {{ manufacturer.country }}</li>
+    {% endfor %}
+  </ul>
```
Comment (TPolina) [inline templates/taxi/manufacturer_list.html:11]: add phrase if there are no manufacturers

**Review by TPolina — CHANGES_REQUESTED** [review]

File: templates/taxi/driver_detail.html [hunk 13]
```diff
@@ -1,8 +1,10 @@
 {% extends "base.html" %}
 {% block content %}
-  <p><strong>First name: </strong>{{ driver.first_name }}, <strong>Last name: </strong>{{ driver.last_name }},
+  <p><strong>Username: </strong> {{ driver.username }}, <strong>First name: </strong>{{ driver.first_name }},
+    <strong>Last name: </strong>{{ driver.last_name }},
   <strong>License number: </strong>{{ driver.license_number }}
   </p>
+  <p><strong>Cars: </strong>{{ driver.car }}</p>
   {% for car in driver.cars.all %}
```
Comment (TPolina) [inline templates/taxi/driver_detail.html:8]: It is better to make info about each car as a list item. Also, add a phrase if there are no cars

File: templates/taxi/driver_list.html [hunk 14]
```diff
@@ -10,6 +10,7 @@ <h1>Driver list</h1>
         <strong>Last name:</strong> {{ driver.last_name }}
       </li>
     {% endfor %}
-    </ul>
```
Comment (TPolina) [inline templates/taxi/driver_list.html]: Why did you delete closing `ul` tag?

File: templates/taxi/car_detail.html [hunk 15]
```diff
@@ -1,10 +1,15 @@
 {% extends "base.html" %}
 {% block content %}
   <p><strong>Car manufacturer: </strong>{{ car.manufacturer.name }}, {{ car.manufacturer.country }}</p>
+  <p><strong>Drivers: </strong></p>
   <hr>
   <ul style="list-style-type:none;">
+  {% if car.drivers %}
   {% for driver in car.drivers.all %}
     <li>First name: {{ driver.first_name }}, Last name: {{ driver.last_name }}</li>
   {% endfor %}
```
Comment (TPolina) [inline templates/taxi/car_detail.html:10]: opening and closing `ul` tags should be around `for` block, `if` and `else` blocks should not be inside it

**Review by TPolina — CHANGES_REQUESTED** [review]

File: templates/taxi/driver_list.html (hunk 14 above)
Comment (kateryna3k) [reply templates/taxi/driver_list.html]: By accident :( Thanks

**Review by MasterpieceElbow — APPROVED** [review]
